File: petclinic-backend-ts/src/owners/dto/owner.dto.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import { IsDefined, IsOptional, Length, Matches, Min, ValidateNested } from 'class-validator';
import { PetDto } from '../../pets/dto/pet.dto';

/**
 * Data transfer object for an owner.
 */
export class OwnerDto {
  @IsOptional()
  @Min(0, { message: 'must be greater than or equal to 0' })
  @ApiPropertyOptional({ readOnly: true, example: 1, description: 'The ID of the pet owner.' })
  id?: number;

  @IsDefined({ message: 'must not be null' })
  @Length(1, 30, { message: 'size must be between 1 and 30' })
  @ApiProperty({ example: 'George', description: 'The first name of the pet owner.' })
  firstName!: string;

  @IsDefined({ message: 'must not be null' })
  @Length(1, 30, { message: 'size must be between 1 and 30' })
  @ApiProperty({ example: 'Franklin', description: 'The last name of the pet owner.' })
  lastName!: string;

  @IsDefined({ message: 'must not be null' })
  @Length(1, 255, { message: 'size must be between 1 and 255' })
  @ApiProperty({ example: '110 W. Liberty St.', description: 'The postal address of the pet owner.' })
  address!: string;

  @IsDefined({ message: 'must not be null' })
  @Length(1, 80, { message: 'size must be between 1 and 80' })
  @ApiProperty({ example: 'Madison', description: 'The city of the pet owner.' })
  city!: string;

  // class-validator emits constraints in reverse declaration order, so declare Length before
  // Matches to emit the pattern message before the size message for telephone.
  @IsDefined({ message: 'must not be null' })
  @Length(1, 20, { message: 'size must be between 1 and 20' })
  @Matches(/^[0-9]*$/, { message: 'must match "^[0-9]*$"' })
  @ApiProperty({ example: '6085551023', description: 'The telephone number of the pet owner.' })
  telephone!: string;

  @ValidateNested({ each: true })
  @Type(() => PetDto)
  @ApiProperty({
    type: () => [PetDto],
    readOnly: true,
    description: 'The pets owned by this individual including any booked vet visits.',
    required: true,
  })
  pets: PetDto[] = [];
}

File: petclinic-backend/src/main/java/victor/training/petclinic/rest/dto/OwnerDto.java
package victor.training.petclinic.rest.dto;

import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Min;
import org.springframework.lang.Nullable;

import java.util.ArrayList;
import java.util.List;

public class OwnerDto extends OwnerFieldsDto {
    @Min(0)
    @Nullable
    @Schema(accessMode = Schema.AccessMode.READ_ONLY, example = "1", description = "The ID of the pet owner.")
    private Integer id;

    @Valid
    @Schema(accessMode = Schema.AccessMode.READ_ONLY,
            description = "The pets owned by this individual including any booked vet visits.",
            requiredMode = Schema.RequiredMode.REQUIRED)
    private List<PetDto> pets = new ArrayList<>();

    public OwnerDto addPetsItem(PetDto petsItem) {
        pets.add(petsItem);
        return this;
    }

    public Integer getId() {
        return id;
    }

    public OwnerDto setId(Integer id) {
        this.id = id;
        return this;
    }

    public List<PetDto> getPets() {
        return pets;
    }

    public OwnerDto setPets(List<PetDto> pets) {
        this.pets = pets;
        return this;
    }

    // Covariant overrides: keep the OwnerDto type while chaining the inherited setters.
    @Override
    public OwnerDto setFirstName(String firstName) {
        super.setFirstName(firstName);
        return this;
    }

    @Override
    public OwnerDto setLastName(String lastName) {
        super.setLastName(lastName);
        return this;
    }

    @Override
    public OwnerDto setAddress(String address) {
        super.setAddress(address);
        return this;
    }

    @Override
    public OwnerDto setCity(String city) {
        super.setCity(city);
        return this;
    }

    @Override
    public OwnerDto setTelephone(String telephone) {
        super.setTelephone(telephone);
        return this;
    }
}
